src/Language/PureScript/TypeChecker/Kinds.hs

```diff
@@ -48,177 +46,213 @@ import Language.PureScript.Names
 import Language.PureScript.TypeChecker.Monad
 import Language.PureScript.Types
 
-instance Partial Kind where
-  unknown = KUnknown
-  isUnknown (KUnknown u) = Just u
-  isUnknown _ = Nothing
-  unknowns = everythingOnKinds (++) go
-    where
-    go (KUnknown u) = [u]
-    go _ = []
-  ($?) sub = everywhereOnKinds go
-    where
-    go t@(KUnknown u) = fromMaybe t $ H.lookup u (runSubstitution sub)
-    go other = other
-
-instance Unifiable Check Kind where
-  KUnknown u1 =?= KUnknown u2 | u1 == u2 = return ()
-  KUnknown u =?= k = u =:= k
-  k =?= KUnknown u = u =:= k
-  Star =?= Star = return ()
-  Bang =?= Bang = return ()
-  Row k1 =?= Row k2 = k1 =?= k2
-  FunKind k1 k2 =?= FunKind k3 k4 = do
-    k1 =?= k3
-    k2 =?= k4
-  k1 =?= k2 = UnifyT . lift . throwError . errorMessage $ KindsDoNotUnify k1 k2
+-- | Generate a fresh kind variable
+freshKind :: (MonadState CheckState m) => m Kind
+freshKind = do
+  k <- gets checkNextKind
+  modify $ \st -> st { checkNextKind = k + 1 }
+  return $ KUnknown k
 
--- |
--- Infer the kind of a single type
---
-kindOf :: Type -> Check Kind
+-- | Update the substitution to solve a kind constraint
+solveKind :: (MonadError MultipleErrors m, MonadState CheckState m) => Int -> Kind -> m ()
+solveKind u k = do
+  occursCheck u k
+  modify $ \cs -> cs { checkSubstitution =
+                         (checkSubstitution cs) { substKind =
+                                                    M.insert u k $ substKind $ checkSubstitution cs
+                                                }
+                     }
+
+-- | Apply a substitution to a kind
+substituteKind :: Substitution -> Kind -> Kind
+substituteKind sub = everywhereOnKinds go
+  where
+  go (KUnknown u) =
+    case M.lookup u (substKind sub) of
+      Nothing -> KUnknown u
+      Just (KUnknown u1) | u1 == u -> KUnknown u1
+      Just t -> substituteKind sub t
+  go other = other
+
+-- | Make sure that an unknown does not occur in a kind
+occursCheck :: (Functor m, Applicative m, MonadError MultipleErrors m) => Int -> Kind -> m ()
+occursCheck _ KUnknown{} = return ()
+occursCheck u k = void $ everywhereOnKindsM go k
+  where
+  go (KUnknown u') | u == u' = throwError . errorMessage . InfiniteKind $ k
+  go other = return other
+
+-- | Unify two kinds
+unifyKinds :: (MonadError MultipleErrors m, MonadState CheckState m) => Kind -> Kind -> m ()
+unifyKinds k1 k2 = do
+  sub <- gets checkSubstitution
+  go (substituteKind sub k1) (substituteKind sub k2)
+  where
+  go (KUnknown u1) (KUnknown u2) | u1 == u2 = return ()
+  go (KUnknown u) k = solveKind u k
+  go k (KUnknown u) = solveKind u k
+  go Star Star = return ()
+  go Bang Bang = return ()
+  go (Row k1') (Row k2') = go k1' k2'
+  go (FunKind k1' k2') (FunKind k3 k4) = do
```

Comment:
This is a perfect example of where shadowing is a good thing, and can prevent you from accidentally referencing stale parameters; i.e. you can bind `k1` and `k2` to make sure you cannot reference them from inside the definition of the function. Albeit in this case, I manually verified the code so it looks good to me. :+1: